{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"background_save":true,"base_uri":"https://localhost:8080/"},"id":"PlO9QphuY1l_","outputId":"f6713495-0328-48af-89a7-08b885ed2a3a"},"outputs":[{"name":"stdout","output_type":"stream","text":["Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (2.0.3)\n","Requirement already satisfied: mlxtend in /usr/local/lib/python3.10/dist-packages (0.22.0)\n","Collecting ipdb\n","  Downloading ipdb-0.13.13-py3-none-any.whl (12 kB)\n","Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2023.4)\n","Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.1)\n","Requirement already satisfied: numpy>=1.21.0 in /usr/local/lib/python3.10/dist-packages (from pandas) (1.25.2)\n","Requirement already satisfied: scipy>=1.2.1 in /usr/local/lib/python3.10/dist-packages (from mlxtend) (1.11.4)\n","Requirement already satisfied: scikit-learn>=1.0.2 in /usr/local/lib/python3.10/dist-packages (from mlxtend) (1.2.2)\n","Requirement already satisfied: matplotlib>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from mlxtend) (3.7.1)\n","Requirement already satisfied: joblib>=0.13.2 in /usr/local/lib/python3.10/dist-packages (from mlxtend) (1.4.2)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from mlxtend) (67.7.2)\n","Requirement already satisfied: ipython>=7.31.1 in /usr/local/lib/python3.10/dist-packages (from ipdb) (7.34.0)\n","Requirement already satisfied: tomli in /usr/local/lib/python3.10/dist-packages (from ipdb) (2.0.1)\n","Requirement already satisfied: decorator in /usr/local/lib/python3.10/dist-packages (from ipdb) (4.4.2)\n","Collecting jedi>=0.16 (from ipython>=7.31.1->ipdb)\n","  Downloading jedi-0.19.1-py2.py3-none-any.whl (1.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m11.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: pickleshare in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (0.7.5)\n","Requirement already satisfied: traitlets>=4.2 in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (5.7.1)\n","Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (3.0.43)\n","Requirement already satisfied: pygments in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (2.16.1)\n","Requirement already satisfied: backcall in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (0.2.0)\n","Requirement already satisfied: matplotlib-inline in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (0.1.7)\n","Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.10/dist-packages (from ipython>=7.31.1->ipdb) (4.9.0)\n","Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (1.2.1)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (0.12.1)\n","Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (4.51.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (1.4.5)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (24.0)\n","Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (9.4.0)\n","Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.0.0->mlxtend) (3.1.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n","Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.0.2->mlxtend) (3.5.0)\n","Requirement already satisfied: parso<0.9.0,>=0.8.3 in /usr/local/lib/python3.10/dist-packages (from jedi>=0.16->ipython>=7.31.1->ipdb) (0.8.4)\n","Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.10/dist-packages (from pexpect>4.3->ipython>=7.31.1->ipdb) (0.7.0)\n","Requirement already satisfied: wcwidth in /usr/local/lib/python3.10/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=7.31.1->ipdb) (0.2.13)\n","Installing collected packages: jedi, ipdb\n","Successfully installed ipdb-0.13.13 jedi-0.19.1\n","Automatic pdb calling has been turned ON\n","    support                                           itemsets\n","0  0.976608  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","1  0.758285  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...\n","2  0.841131  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...\n","3  0.571150  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","4  0.916179  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","                                         antecedents  \\\n","0  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","1  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...   \n","2  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","3  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","4  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","\n","                                         consequents  antecedent support  \\\n","0  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...            0.976608   \n","1  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.841131   \n","2  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.571150   \n","3  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.976608   \n","4  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.916179   \n","\n","   consequent support   support  confidence      lift  leverage  conviction  \\\n","0            0.841131  0.822612    0.842315  1.001409  0.001157    1.007514   \n","1            0.976608  0.822612    0.977984  1.001409  0.001157    1.062481   \n","2            0.976608  0.565302    0.989761  1.013468  0.007512    2.284600   \n","3            0.571150  0.565302    0.578842  1.013468  0.007512    1.018264   \n","4            0.976608  0.916179    1.000000  1.023952  0.021431         inf   \n","\n","   zhangs_metric  \n","0       0.060130  \n","1       0.008854  \n","2       0.030987  \n","3       0.568103  \n","4       0.279070  \n"]}],"source":["!pip install pandas mlxtend ipdb\n","import pandas as pd\n","from mlxtend.frequent_patterns import apriori, association_rules\n","\n","%pdb on\n","\n","#Load the Excel file\n","file_path = 'IBS Dataset.xlsx'\n","xls = pd.ExcelFile(file_path)\n","\n","# Load the abundance data\n","abundance = pd.read_excel(xls, 'abundance')\n","abundance.set_index('sample-id', inplace=True)\n","\n","# Normalize the abundance data\n","abundance_normalized = abundance.div(abundance.sum(axis=1), axis=0)\n","\n","# Binarize the data and ensure it's boolean type\n","abundance_binary = abundance_normalized.gt(0)\n","\n","# Generate frequent itemsets\n","frequent_itemsets = apriori(abundance_binary, min_support=0.5,max_len=None, use_colnames=True,verbose=0, low_memory= True)\n","# Generate association rules\n","rules = association_rules(frequent_itemsets, metric=\"lift\", min_threshold=1.0)\n","\n","# Display the frequent itemsets and association rules\n","print(frequent_itemsets.head())\n","print(rules.head())\n","\n"]},{"cell_type":"markdown","source":["-------------------------------------------------------------------------------\n","--------------------------------------------------------------------------------\n","\n","The output of the association analysis provides insights into the co-occurrence patterns of bacterial taxa in the context of Irritable Bowel Syndrome (IBS).\n","It reveals relationships and associations between different bacteria that are prevalent in the samples.\n","\n","\n","### Frequent Itemsets\n","\n","The frequent itemsets indicate combinations of bacterial taxa that are commonly found together in the samples. For example:\n","```\n","| support | itemsets                                             |\n","|---------|------------------------------------------------------|\n","| 0.976608| (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...)  |\n","```\n","This itemset, with a high support value (97.66%), suggests that these bacterial taxa are almost always present together in the samples. This high prevalence could indicate a stable core microbiota that might be typical in individuals with IBS or could be part of a healthy microbiome.\n","\n","### Association Rules\n","\n","The association rules provide more detailed relationships between bacterial taxa. For example:\n","```\n","| antecedents                                      | consequents                                       | support  | confidence | lift    | leverage | conviction | zhangs_metric |\n","|--------------------------------------------------|--------------------------------------------------|----------|------------|---------|----------|------------|---------------|\n","| (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...) | (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...) | 0.822612 | 0.842315   | 1.001409| 0.001157 | 1.007514   | 0.060130      |\n","```\n","This rule indicates that when the taxa from `Bacteroidota` are present, there's an 84.23% confidence that the taxa from `Firmicutes` will also be present. The lift value of 1.001409 suggests that this co-occurrence is slightly more likely than random chance.\n","\n","### Interpretation in the Context of IBS\n","\n","1. **Microbiota Composition**:\n","   - **High Support Itemsets**: The bacterial taxa that frequently co-occur might be part of the typical gut microbiota of individuals with IBS. High support values suggest these taxa are common across many samples.\n","   - **Core Microbiota**: The frequent presence of certain itemsets may indicate a core group of bacteria that are essential for gut health or are consistently altered in IBS patients.\n","\n","2. **Potential Dysbiosis**:\n","   - **Association Rules**: The rules highlight specific relationships between bacterial groups. Changes in these relationships (compared to healthy individuals) could indicate dysbiosis (imbalance in the microbial community) associated with IBS.\n","   - **Altered Associations**: If certain bacterial taxa that usually do not co-occur in healthy individuals are found together in IBS samples, it might suggest a condition-specific alteration in the microbiota.\n","\n","3. **Bacterial Interactions**:\n","   - The confidence and lift values help identify strong associations between bacterial taxa. Understanding these associations can provide insights into how bacterial communities interact within the gut environment of IBS patients.\n","\n","### Limitations and Further Research\n","\n","- **Causation vs. Correlation**: The analysis shows associations but does not prove causation. Further experimental studies are needed to identify causative relationships.\n","- **Comparative Analysis**: Comparing these results with microbiota data from healthy individuals could help identify specific bacterial changes linked to IBS.\n","- **Functional Insights**: Metagenomic or metabolomic studies could provide functional insights into how these bacterial taxa contribute to IBS symptoms.\n","\n","In conclusion, the output indicates common bacterial associations in the IBS dataset, which can help in understanding the microbial ecology in IBS.\n","\n","-------------------------------------------------------------------------------\n","--------------------------------------------------------------------------------"],"metadata":{"id":"3jx0GdteaVEp"}},{"cell_type":"code","execution_count":2,"metadata":{"id":"I3t4iI3lfbRH","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1717088933721,"user_tz":-330,"elapsed":165303,"user":{"displayName":"Suryansh Rm","userId":"14379116401228697465"}},"outputId":"72f04a51-9668-4744-99c1-75e2f90e2ffb"},"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n","  and should_run_async(code)\n"]},{"output_type":"stream","name":"stdout","text":["Significant differences:\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides: p-value = 0.012641737928569083\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter: p-value = 1.971742366336403e-05\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Prevotellaceae;g__Prevotella_9: p-value = 0.03271102997242126\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium: p-value = 0.02196965542552386\n","d__Bacteria;p__Actinobacteriota;c__Actinobacteria;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium: p-value = 0.01262786342752088\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__uncultured: p-value = 0.00011330337908593914\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Fusicatenibacter: p-value = 0.027666583521733006\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Ruminococcus]_torques_group: p-value = 0.013278892976305473\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__UCG-002: p-value = 0.04624921975501301\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus: p-value = 0.0005701350797954617\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;__;__: p-value = 0.025641900941615316\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Collinsella: p-value = 6.39071684290037e-06\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Roseburia: p-value = 0.011542891994344554\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Coprococcus: p-value = 0.007021576757395204\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__[Eubacterium]_coprostanoligenes_group;g__[Eubacterium]_coprostanoligenes_group: p-value = 0.00014267596197919393\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__: p-value = 0.0022515263889038317\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;__: p-value = 0.0040540887665470576\n","d__Bacteria;p__Firmicutes;c__Negativicutes;o__Acidaminococcales;f__Acidaminococcaceae;g__Phascolarctobacterium: p-value = 0.048457157310964866\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Monoglobales;f__Monoglobaceae;g__Monoglobus: p-value = 0.0033572513141230747\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;__: p-value = 7.71718763937616e-06\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Eubacterium]_ventriosum_group: p-value = 0.0014449167465662807\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;__: p-value = 0.0018680725224445754\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Incertae_Sedis: p-value = 0.00015612564593618143\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Ruminococcus]_gauvreauii_group: p-value = 0.009931019463805875\n","d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacterales;f__Pectobacteriaceae;g__Dickeya: p-value = 6.88838178973189e-05\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Turicibacter: p-value = 7.119720694580543e-05\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Prevotellaceae;g__Prevotella: p-value = 0.0031379047850195554\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Butyricicoccaceae;g__Butyricicoccus: p-value = 0.0033934009997090364\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__Flavonifractor: p-value = 2.555471936549293e-11\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Lachnospiraceae_UCG-001: p-value = 0.0008057958223600243\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__Intestinimonas: p-value = 0.000175647328679515\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Eggerthella: p-value = 0.007562198649020495\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;g__Family_XIII_AD3011_group: p-value = 2.173909248267497e-05\n","d__Bacteria;p__Firmicutes;c__Negativicutes;o__Veillonellales-Selenomonadales;f__Veillonellaceae;g__Veillonella: p-value = 1.2209454293132948e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Clostridia_UCG-014;f__Clostridia_UCG-014;g__Clostridia_UCG-014: p-value = 0.008298480532199769\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__[Eubacterium]_siraeum_group: p-value = 0.005474031686590666\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Eubacterium]_eligens_group: p-value = 0.008397369241016183\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__UCG-005: p-value = 0.002780851016108903\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Eubacterium]_xylanophilum_group: p-value = 0.03438654634447196\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Enterococcaceae;__: p-value = 0.006973520153345907\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriales_Incertae_Sedis;g__uncultured: p-value = 0.03999626134753967\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__Oscillibacter: p-value = 0.00033327763986226423\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Muribaculaceae: p-value = 0.0007958729041106978\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__[Clostridium]_innocuum_group: p-value = 1.5227087876379948e-08\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;g__[Eubacterium]_brachy_group: p-value = 8.339917682482592e-06\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Lachnospiraceae_UCG-010: p-value = 0.006281046198307063\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__UBA1819: p-value = 1.2232862144446294e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Lachnospiraceae_NK4A136_group: p-value = 0.0024543539593231307\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Lactococcus: p-value = 0.00999508350754495\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelatoclostridiaceae;g__Erysipelatoclostridium: p-value = 4.112715480344582e-07\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Faecalitalea: p-value = 0.0014019012140354065\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Carnobacteriaceae;g__Granulicatella: p-value = 0.0022350469122335874\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__Romboutsia: p-value = 0.0032927231339959858\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lactobacillus: p-value = 0.019553423596789545\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;__: p-value = 0.0040305877112240115\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;g__Family_XIII_UCG-001: p-value = 0.024148145741181362\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__Terrisporobacter: p-value = 0.005798730306951396\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Negativibacillus: p-value = 0.0015626228271006265\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;g__[Eubacterium]_nodatum_group: p-value = 2.4602084696275594e-07\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Ligilactobacillus: p-value = 0.003946613453560563\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Paludicola: p-value = 3.1632362056216636e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;__: p-value = 0.004129948656427291\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Anaerotruncus: p-value = 0.02315200367699634\n","d__Bacteria;p__Firmicutes;c__Clostridia;__;__;__: p-value = 0.0009775740168948897\n","d__Bacteria;p__Cyanobacteria;c__Cyanobacteriia;o__Chloroplast;f__Chloroplast;g__Chloroplast: p-value = 0.02437755856442082\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Porphyromonadaceae;g__Porphyromonas: p-value = 0.009557246717601827\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Solobacterium: p-value = 0.003549428074399871\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__Peptostreptococcaceae: p-value = 0.00375394832423253\n","d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodobacterales;f__Rhodobacteraceae;__: p-value = 0.0015099349126481915\n","d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia: p-value = 0.02183131115959037\n","d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Devosiaceae;g__Maritalea: p-value = 0.02163342375816425\n","d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia-Shigella: p-value = 0.0005982812480771056\n","Unassigned;__;__;__;__;__: p-value = 1.2475073533371e-07\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Slackia: p-value = 0.02514006048690314\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostignum: p-value = 0.0031621683844138184\n","d__Bacteria;p__Actinobacteriota;c__Actinobacteria;o__Micrococcales;f__Micrococcaceae;g__Rothia: p-value = 1.290491102284036e-07\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Family_XI;g__Finegoldia: p-value = 0.006300126709893108\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Atopobiaceae;g__Atopobium: p-value = 7.182428618338719e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Family_XI;g__Parvimonas: p-value = 1.2871956889135136e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Candidatus_Soleaferrea: p-value = 0.018546795417376732\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Gordonibacter: p-value = 0.00099004802776151\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__uncultured;g__uncultured: p-value = 0.04415378888223993\n","d__Bacteria;__;__;__;__;__: p-value = 5.109933563421111e-12\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__Intestinibacter: p-value = 0.02453462289558687\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Prevotellaceae;g__Prevotellaceae_UCG-001: p-value = 0.008883437156919338\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Holdemania: p-value = 7.713554177229272e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__[Clostridium]_methylpentosum_group;g__[Clostridium]_methylpentosum_group: p-value = 0.026495963661786845\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Adlercreutzia: p-value = 0.0013570974181515202\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Anaerovoracaceae;g__Mogibacterium: p-value = 0.0001884399854189884\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__[Eubacterium]_fissicatena_group: p-value = 0.01723719503000666\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Lachnospiraceae_NK4B4_group: p-value = 0.031097374743878883\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Peptostreptococcales-Tissierellales;f__Peptostreptococcaceae;g__Clostridioides: p-value = 0.016703741440464808\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriales_Incertae_Sedis;g__Raoultibacter: p-value = 0.0008487699440209789\n","d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae;g__Pseudomonas: p-value = 9.18504821684795e-09\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospirales;g__Hydrogenoanaerobacterium: p-value = 0.0177032467168689\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelatoclostridiaceae;g__Coprobacillus: p-value = 0.03387312737876455\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Christensenellales;f__Christensenellaceae;g__uncultured: p-value = 0.025622832859803404\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__CHKCI002: p-value = 0.022849238412734858\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Eubacteriales;f__Anaerofustaceae;g__Anaerofustis: p-value = 0.028065090317332436\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Dielma: p-value = 0.0047433658582259975\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__CAG-873: p-value = 0.010741313527748663\n","d__Bacteria;p__Cyanobacteria;c__Vampirivibrionia;o__Gastranaerophilales;f__Gastranaerophilales;g__Gastranaerophilales: p-value = 0.022931959147611286\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Limosilactobacillus: p-value = 0.0005243350762309552\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Prevotellaceae;g__Prevotellaceae_UCG-003: p-value = 0.0005818347859397863\n","d__Bacteria;p__Firmicutes;__;__;__;__: p-value = 0.04595802482492773\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__DTU089: p-value = 0.03343516597126014\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Christensenellales;f__Christensenellaceae;g__Catabacter: p-value = 0.012641737928568923\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Oribacterium: p-value = 0.00900729179008184\n","d__Bacteria;p__Firmicutes;c__Bacilli;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Merdibacter: p-value = 0.0024781726660260735\n","d__Bacteria;p__Actinobacteriota;c__Actinobacteria;o__Actinomycetales;f__Actinomycetaceae;g__Actinomyces: p-value = 3.57182499078356e-06\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Barnesiellaceae;g__uncultured: p-value = 0.01377978353385338\n","d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Burkholderiales;f__Oxalobacteraceae;__: p-value = 0.020108736083525598\n","d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Methanomassiliicoccales;f__Methanomassiliicoccaceae;g__Methanomassiliicoccus: p-value = 0.005650618528667944\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Sphingobacteriales;f__Sphingobacteriaceae;g__Pedobacter: p-value = 8.115596836402097e-05\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Muribaculum: p-value = 5.721508282813313e-05\n","d__Bacteria;p__Actinobacteriota;c__Actinobacteria;o__Corynebacteriales;f__Corynebacteriaceae;g__Corynebacterium: p-value = 2.2180872636665745e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Cuneatibacter: p-value = 7.059259481239333e-05\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Oscillospiraceae;g__Sporobacter: p-value = 0.04436115202249166\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__CHKCI001: p-value = 0.03657958624487464\n","d__Bacteria;p__Actinobacteriota;c__Actinobacteria;o__Actinomycetales;f__Actinomycetaceae;g__F0332: p-value = 0.0011366794107957635\n","d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Dysgonomonadaceae;g__Dysgonomonas: p-value = 0.040063769316399274\n","d__Bacteria;p__Fusobacteriota;c__Fusobacteriia;o__Fusobacteriales;f__Leptotrichiaceae;g__Sneathia: p-value = 0.028777663642128328\n","d__Bacteria;p__Firmicutes;c__Clostridia;o__Oscillospirales;f__Ethanoligenenaceae;g__Acetanaerobacterium: p-value = 0.0028737598195786116\n","d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Enorma: p-value = 0.04307682874338716\n","    support                                           itemsets\n","0  0.976608  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","1  0.758285  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...\n","2  0.841131  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...\n","3  0.571150  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","4  0.916179  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...\n","                                         antecedents  \\\n","0  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...   \n","1  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","2  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","3  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","4  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...   \n","\n","                                         consequents  antecedent support  \\\n","0  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.841131   \n","1  (d__Bacteria;p__Firmicutes;c__Clostridia;o__La...            0.976608   \n","2  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.571150   \n","3  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.976608   \n","4  (d__Bacteria;p__Bacteroidota;c__Bacteroidia;o_...            0.916179   \n","\n","   consequent support   support  confidence      lift  leverage  conviction  \\\n","0            0.976608  0.822612    0.977984  1.001409  0.001157    1.062481   \n","1            0.841131  0.822612    0.842315  1.001409  0.001157    1.007514   \n","2            0.976608  0.565302    0.989761  1.013468  0.007512    2.284600   \n","3            0.571150  0.565302    0.578842  1.013468  0.007512    1.018264   \n","4            0.976608  0.916179    1.000000  1.023952  0.021431         inf   \n","\n","   zhangs_metric  \n","0       0.008854  \n","1       0.060130  \n","2       0.030987  \n","3       0.568103  \n","4       0.279070  \n"]}],"source":["# Load the Excel file\n","file_path = 'IBS Dataset.xlsx'\n","xls = pd.ExcelFile(file_path)\n","\n","# Load the abundance data\n","abundance = pd.read_excel(xls, 'abundance')\n","abundance.set_index('sample-id', inplace=True)\n","\n","# Normalize the abundance data\n","abundance_normalized = abundance.div(abundance.sum(axis=1), axis=0)\n","\n","# Binarize the data and ensure it's boolean type\n","abundance_binary = abundance_normalized.gt(0)\n","\n","# Load the metadata sheet\n","metadata = pd.read_excel(xls, 'sample_ids_and_sentences')\n","metadata.set_index('sample-id', inplace=True)\n","\n","# Merge abundance data with metadata\n","merged_data = pd.merge(abundance_binary, metadata, left_index=True, right_index=True)\n","\n","# Perform group comparison using t-tests or any other statistical tests\n","ibs_group = merged_data[merged_data['Group'] == 'IBS']\n","control_group = merged_data[merged_data['Group'] == 'Control']\n","\n","# You can replace these columns with actual column names\n","microbial_columns = abundance_binary.columns\n","\n","# For example, using t-tests\n","from scipy.stats import ttest_ind\n","\n","significant_differences = {}\n","for column in microbial_columns:\n","    t_stat, p_value = ttest_ind(ibs_group[column], control_group[column])\n","    if p_value < 0.05:\n","        significant_differences[column] = p_value\n","\n","# Print significant differences\n","print(\"Significant differences:\")\n","for bacteria, p_value in significant_differences.items():\n","    print(f\"{bacteria}: p-value = {p_value}\")\n","\n","# Generate frequent itemsets\n","frequent_itemsets = apriori(abundance_binary, min_support=0.5, max_len=None, use_colnames=True, verbose=0, low_memory=True)\n","\n","# Generate association rules\n","rules = association_rules(frequent_itemsets, metric=\"lift\", min_threshold=1.0)\n","\n","# Display the frequent itemsets and association rules\n","print(frequent_itemsets.head())\n","print(rules.head())\n"]},{"cell_type":"markdown","source":["--------------------------------------------------------------------------------------------------------------------------------------------------------------\n","\n","The output identifies specific bacterial taxa with significantly different abundances between IBS patients and healthy controls, highlighting potential biomarkers for IBS.\n"," Genera like `Agathobacter` and `Butyricicoccus` show highly significant differences, suggesting their crucial role in the gut microbiome's link to IBS. These findings can inform targeted therapies, such as probiotics, and aid in understanding IBS pathogenesis. Further validation and functional analysis of these taxa can enhance personalized treatment strategies for IBS.\n","\n"," --------------------------------------------------------------------------------------------------------------------------------------------------------------\n"],"metadata":{"id":"LE7oRWuXfNUi"}},{"cell_type":"code","source":[],"metadata":{"id":"E2onzLOhfRAO"},"execution_count":null,"outputs":[]}],"metadata":{"accelerator":"GPU","colab":{"gpuType":"T4","provenance":[],"authorship_tag":"ABX9TyP78noZyEaLCsUbHziNBCSP"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}